contributors/design-proposals/network/grpc-cni-plugin.md

```diff
@@ -0,0 +1,268 @@
+
+The `RemoteCNINetworkPlugin` object will include a gRPC client named `cniCLient`, the kubelet invokes for `RemoteCNINetworkPlugin` will be convert to corresponding RPCs calls by `cniCLient`.
+
+```go
+// RemoteCNINetworkPlugin is a GRPC implementation of NetworkPlugin
+type RemoteCNINetworkPlugin struct {
+
+
+  // host is an interface that plugin can use to access the kubelet
+  // plugin will use host to get network namespace information
+  // of sandbox
+  host      network.Host
+
+  // cniClient is the client for CNIDriver service
+  // plugin will convert kubelet call to invokes of this client
+  cniClient cniClient
+}
+
+func (plugin *RemoteCNINetworkPlugin) SetUpPod{
+  // *1 Get sandbox network namespace
+  plugin.host.GetNetNS
+
+  // *2 Get portMap
+  plugin.host.GetPodPortMapping
+
+  // *3 build runtime config
+
+  // *4 Call cniClient
+  // cniClient.setUpPod will call DaemonSet CNI driver
+  // to add CNI network
+  plugin.cniClient.setUpPod
+}
+
+func (plugin *RemoteCNINetworkPlugin) TearDownPod{
+  // *1 Get sandbox network namespace
+  plugiun.host.GetNetNS
+
+  // *2 Get portMap
+  plugin.host.GetPodPortMapping
+
+  // *3 build runtime config
+
+  // *4 Call cniClient
+  // cniClient.tearDownPod will call DaemonSet CNI driver
+  // to delete CNI network
+  plugin.cniClient.tearDownPod
+}
+
+func (plugin *RemoteCNINetworkPlugin) Status{
+  // call cniClient
+  // cniCLient.cniStatus will call DaemonSet CNI driver
+  // to sync CNI network config and check whether
+  // CNI driver initialized
+  plugin.cniClient.cniStatus
+}
+```
+
+#### DaemonSet CNI driver program
+
+We will build a new repo for out-tree CNI driver, maybe named `kubernetes-cni`, which will run as a DeamonSet program,  and offer a gRPC server that in-tree network plugin can invoke.  This CNI driver will acts as a bridge between third-party CNI plugin binaries and kubelet.
+
+And the DaemonSet CNI driver will use a configmap as the network config file, so don’t have to access the root filesystem of the node or master.
+
+```go
+type cniDriver struct {
+  // cni.NetworkConfig is the network config specified by
+  // containernetworking/cni
+  // CNI driver will build this config according to network
+  // information from configmap
+  networkConfig cni.NetworkConfig
+
+  // cni.RuntimeConf is the runtime config specified by
+  // containernetworking/cni
+  // CNI driver will build this config according to runtime config
+  // from kubelet
+  runtimeConfig cni.RuntimeConf
+
+  // cni.CNI is the interface specified by containernetworking/cni
+  // cniDriver will use this interface to Add/Del network
+  cniConfig     cni.CNI
+
+  // cniPluginDir is the path of third-party CNI plugin binaries
+  // which from environment variables of DaemonSet
+  cniPluginDir     []string
+
+  // networkConfigDir is the network config path
+  // which from environment variables of DaemonSet
+  networkConfigDir string

```

Comment:
I'm a bit confused - what is this parameter used for? I thought the network config was specified in a config map.

Is this just to configure the driver to know where the configmap has been mounted into the pod?